### ISSUE TITLE
[uwebsockets] version 20.44.0 is unusable

### ISSUE THREAD

**Issue by nbkolchin** [microsoft/vcpkg#31813]
**Describe the bug**
Some versions of uWebSockets require untagged uSockets version (they are using git modules). Blindly updating port tags will result in build failures when use application tries to use uWebSockets, because uSockets API has different and uWebSockets is a header only library.

This was reported several times to upstream:

- https://github.com/uNetworking/uWebSockets/issues/1615
- https://github.com/uNetworking/uWebSockets/issues/1545

but still happens sometime.

**Environment**
- OS: any
- Compiler: any

**To Reproduce**
Try to build any application using uWebSockets with vcpkg baseline 662dbb50e63af15baa2909b7eac5b1b87e86a0aa.

**Expected behavior**
Two variants:

- Don't update uWebSockets if it have untagged uSockets dependency.
- Always keep uSockets synced with uWebSockets via git hash.

**Failure logs**
```
...
.build-optimize/vcpkg_installed/x64-linux/include/uwebsockets/App.h:345:79:   required from ‘uWS::TemplatedApp<SSL>&& uWS::TemplatedApp<SSL>::ws(std::string, WebSocketBehavior<UserData>&&) [with UserData = Data; bool SSL = false; std::string = std::__cxx11::basic_string<char>]’
/app/ws/crossx_ws.cpp:71:16:   required from here
.build-optimize/vcpkg_installed/x64-linux/include/uwebsockets/WebSocket.h:113:43: error: ‘us_socket_write2’ was not declared in this scope; did you mean ‘us_socket_write’?
  113 |             int written = us_socket_write2(0, (struct us_socket_t *)this, header, header_length, message.data(), (int) message.length());
...
  ```

**Additional context**

.

**Comment by dg0yt** [comment]
> Blindly updating port tags will result in build failures when user application tries to use uWebSockets, because uSockets has different API and uWebSockets is a header only library.

Can you provide, or point to, a minimal test project, so that it could be considered for a test port in vcpkg CI?

**Comment by uNetworkingAB** [comment]
I've tagged uSockets v0.8.6 now. But there is always a human factor where I can simply miss. So just ping me when there's an issue and I will tag it.

@dg0yt You won't catch these issues with a minimal project since 99% of uSockets is the same; v0.8.6 only adds 1 new function and you won't get linker error until you use that path, which is not guaranteed in a small test project.

**Comment by dg0yt** [comment]
> @dg0yt You won't catch these issues with a minimal project since 99% of uSockets is the same; v0.8.6 only adds 1 new function and you won't get linker error until you use that path, which is not guaranteed in a small test project.

I'm aware of that.

What is meant by "tags" here, i.e. how do tags turn an incompatible version of usockets into a compatible version?

BTW in manifest mode users can easily combine compatible versions. But I assume the UX in arriving at such a combination isn't satisfying.

**Comment by nbkolchin** [comment]
> > Blindly updating port tags will result in build failures when user application tries to use uWebSockets, because uSockets has different API and uWebSockets is a header only library.
> 
> Can you provide, or point to, a minimal test project, so that it could be considered for a test port in vcpkg CI?

While @uNetworkingAB definetely knows more than me on this topic, you can use "builtin" uWebSocket examples for "build testing". It is better to have some testing.

> BTW in manifest mode users can easily combine compatible versions.

How? I haven't found anything in documentation on this.

**Comment by uNetworkingAB** [comment]
a tag is a git tag

**Comment by dg0yt** [comment]
Hm. If I use the word "release" instead of "tag", is this accurate?

uuwebsockets releases may rely on unreleased version of usockets.
usockets doesn't export version information (to cpp macros, CMake config, pkg-config).
Therefore, uwebsocket build or usage cannot check early for a configuration mismatch.
Therefore, port maintainer shall manually check uwebsockets' usocket git submodule before an update.

**Comment by uNetworkingAB** [comment]
> uuwebsockets releases may rely on unreleased version of usockets.

Yes but only from me forgetting to tag a release in uSockets. So if there is no matching tag for whatever the submodule points at, then you can contact me and I will tag it.

The reason I forget is because I don't use package managers to link the two projects, I simply link them by git submodules. 

But to simplify packing them in a package manager, just ping me when you miss a tag and I will tag it.
